services/venice-controller/src/main/java/com/linkedin/venice/controller/systemstore/SystemStoreHealthCheckService.java

```diff
@@ -0,0 +1,164 @@
+import java.time.Duration;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * This class is used to schedule periodic check of user system stores in Venice cluster to make sure system stores has
+ * good version and is ingesting correctly. This service is expected to be setup in the leader controller of the cluster,
+ * and it will periodically scan all user system stores and collect bad system stores. Parent controller will hit the
+ * controller endpoint to collect information about bad system stores.
+ */
+public class SystemStoreHealthCheckService extends AbstractVeniceService {
+  public static final Logger LOGGER = LogManager.getLogger(SystemStoreHealthCheckService.class);
+  private final ReadWriteStoreRepository storeRepository;
+  private final PushStatusStoreReader pushStatusStoreReader;
+  private final PushStatusStoreWriter pushStatusStoreWriter;
+  private final MetaStoreReader metaStoreReader;
+  private final MetaStoreWriter metaStoreWriter;
+
+  private final int checkPeriodInSeconds;
+  private final AtomicBoolean isRunning = new AtomicBoolean(false);
+  private ScheduledExecutorService checkServiceExecutor;
+
+  private Set<String> unhealthySystemStoreSet = new HashSet<>();
+
+  public SystemStoreHealthCheckService(
+      ReadWriteStoreRepository storeRepository,
+      MetaStoreReader metaStoreReader,
+      MetaStoreWriter metaStoreWriter,
+      PushStatusStoreReader pushStatusStoreReader,
+      PushStatusStoreWriter pushStatusStoreWriter,
+      int systemStoreCheckPeriodInSeconds) {
+    this.storeRepository = storeRepository;
+    this.metaStoreWriter = metaStoreWriter;
+    this.metaStoreReader = metaStoreReader;
+    this.pushStatusStoreWriter = pushStatusStoreWriter;
+    this.pushStatusStoreReader = pushStatusStoreReader;
+    this.checkPeriodInSeconds = systemStoreCheckPeriodInSeconds;
+  }
+
+  /**
+   * Return unhealthy system store name set. This API is expected to be called by parent controller.
+   */
+  public Set<String> getUnhealthySystemStoreSet() {
+    return unhealthySystemStoreSet;
+  }
+
+  @Override
+  public boolean startInner() {
+    checkServiceExecutor = Executors.newScheduledThreadPool(1);
+    isRunning.set(true);
+    checkServiceExecutor.scheduleWithFixedDelay(
+        new SystemStoreHealthCheckTask(),
+        checkPeriodInSeconds,
+        checkPeriodInSeconds,
+        TimeUnit.SECONDS);
+    return true;
+  }
+
+  @Override
+  public void stopInner() {
+    isRunning.set(false);
+    checkServiceExecutor.shutdownNow();
+  }
+
+  private class SystemStoreHealthCheckTask implements Runnable {
+    @Override
+    public void run() {
```

Comment:
Nit, could we add one line of log to indicate the service states when it is started/stopped.

Reply:
Yeah, sure, looks like current impl does not have any log, I will add some for critical places for debugging purpose.